src/library/scala/collection/immutable/VectorMap.scala

```diff
@@ -0,0 +1,155 @@
+package scala
+package collection
+package immutable
+
+import java.io.{ObjectInputStream, ObjectOutputStream}
+
+import scala.collection.mutable.{Builder, ImmutableBuilder}
+import scala.annotation.unchecked.{uncheckedVariance => uV}
+import scala.collection.immutable.Map.useBaseline
+
+/** This class implements immutable maps using a vector/map-based data structure, which preserves insertion order.
+  *
+  *  Unlike `ListMap`, `VectorMap` has amortized effectively constant lookup at the expense
+  *  of using extra memory and generally lower performance for other operations
+  *
+  *  @tparam K      the type of the keys contained in this vector map.
+  *  @tparam V      the type of the values associated with the keys in this vector map.
+  *
+  * @author Matthew de Detrich
+  * @version 2.13
+  * @since 2.13
+  * @define coll immutable vector map
+  * @define Coll `immutable.VectorMap`
+  */
+final class VectorMap[K, +V] private[immutable] (
+    val fields: Vector[K],
+    val underlying: Map[K, (Int, V)])
+    extends AbstractMap[K, V]
+    with LinkedMap[K, V]
+    with MapOps[K, V, VectorMap, VectorMap[K, V]]
+    with StrictOptimizedIterableOps[(K, V), Iterable, VectorMap[K, V]] {
+
+  override protected[this] def className: String = "VectorMap"
+
+  override def updated[V1 >: V](key: K, value: V1): VectorMap[K, V1] = {
+    if (underlying.get(key).isDefined) {
+      val oldKey = underlying(key)._1
+      new VectorMap(fields,
+                    underlying.updated(key, (oldKey, value.asInstanceOf[V])))
+    } else {
+      new VectorMap(
+        fields :+ key,
+        underlying.updated(key, (fields.length + 1, value.asInstanceOf[V])))
+    }
+  }
+
+  override def withDefault[V1 >: V](d: K => V1): Map.WithDefault[K, V1] =
+    new Map.WithDefault(this, d)
```

Comment:
I don't believe so, since `Map.WithDefault` wraps an underlying Map (which in this case is going to be a `VectorMap`). It will also use the underlying `updated`/`remove` methods, which means that the underlying `VectorMap` will get replaced with another `VectorMap`.

However there is a point that we do lose the original type, since `withDefault` will be of type `Map` rather than `VectorMap`, not exactly sure what to do here?

Reply:
The behavior looks OK at least, thanks for the investigation!

Reply:
Should write a test to verify this behavior though, will do it tonight

Reply:
Committed a test to verify this behaviour